src/includes/dinit-iostream.h

```diff
@@ -0,0 +1,742 @@
+    }
+
+    ostream(const int tfd) noexcept
+    {
+        fd = tfd;
+        buf = std::unique_ptr<streambuf>(new(std::nothrow) streambuf);
+    }
+
+    ostream(const int tfd, std::unique_ptr<streambuf> passed_buf) noexcept
+    {
+        fd = tfd;
+        buf = std::move(passed_buf);
+    }
+
+    ostream(const ostream &) = delete;
+
+    void operator=(const ostream &) = delete;
+
+    ostream(const ostream &&) = delete;
+
+    void operator=(const ostream &&) = delete;
+
+    // Open file path which specified in path or as open(2) parameter with flags or modes.
+    //
+    // Note: O_WRONLY is hard-coded in internal call to POSIX open(2).
+    //
+    // Note: Opening an already opened stream is not allowed and caller must close the stream
+    // before open it again.
+    //
+    // Returns: true on success, false on failure.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_system_err on POSIX open(2) failure.
+    bool open_nx(const char *path) noexcept;
+    bool open(const char *path);
+
+    bool open_nx(const char *path, const int flags) noexcept;
+    bool open(const char *path, const int flags);
+
+    bool open_nx(const char *path, const int flags, const mode_t mode) noexcept;
+    bool open(const char *path, const int flags, const mode_t mode);
+
+    // Flush the buffer and close the current file descriptor.
+    //
+    // Returns: true on success, false on failure cases.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time).
+    // dio::iostream_system_err on system I/O failure.
+    bool close_nx() noexcept;
+    bool close();
+
+    // Returns: current bits/vars in stream based on io_states enum.
+    int current_state() noexcept;
+
+    // Returns: true when there is no set bits/vars otherwise false.
+    bool good() noexcept;
+
+    // Returns: true when buffer_failbit is set otherwise false.
+    bool buffer_failure() noexcept;
+
+    // Returns: 0 or errno based on io_error value.
+    int io_failure() noexcept;
+
+    // Resets failure bits/vars to false (or 0 about io_error).
+    // Note: buffer_failbit will not be reset.
+    void clear() noexcept;
+
+    // Flush buffer of stream.
+    //
+    // Returns: true on full-flushed buffer, false on failure or not full-flushed buffer.
+    //
+    // Non _nx variant may throw:
+    // dio::iostream_internal_err on Buffer failure (buffer did not allocated at construction time).
+    // dio::iostream_system_err on system I/O failure.
+    bool flush_nx() noexcept;
+    bool flush();
+
+    // write() functions, Used to write many types.
+    //
+    // All calls to write() will be buffered and not necessarily written to the file descriptor except
+    // in these situations:
+    // 1. The size of the passed message is larger than the buffer size.
+    // 2. The size of the passed message is larger than the current buffer free space.
```

Comment:
These (points 1 and 2) are basically implementation detail, so omit them from documentation. It shouldn't matter to the caller.

You could combine them as "1. The buffer becomes full" for example.